Python/ceval_gil.c

```diff
@@ -1192,6 +1192,69 @@ _PyEval_DisableGIL(PyThreadState *tstate)
 }
 #endif
 
+#if defined(Py_REMOTE_DEBUG) && defined(Py_SUPPORTS_REMOTE_DEBUG)
+// Note that this function is inline to avoid creating a PLT entry
+// that would be an easy target for a ROP gadget.
+static inline void run_remote_debugger_script(const char *path)
+{
+    if (0 != PySys_Audit("remote_debugger_script", "s", path)) {
+        PyErr_FormatUnraisable(
+            "Audit hook failed for remote debugger script %s", path);
+        return;
+    }
+
+    // Open the debugger script with the open code hook, and reopen the
+    // resulting file object to get a C FILE* object.
+    PyObject* fileobj = PyFile_OpenCode(path);
+    if (!fileobj) {
+        PyErr_FormatUnraisable("Can't open debugger script %s", path);
+        return;
+    }
+
+    int fd = PyObject_AsFileDescriptor(fileobj);
+    if (fd == -1) {
+        PyErr_FormatUnraisable("Can't find fd for debugger script %s", path);
+    } else {
+        int dup_fd = -1;
+        FILE *f = NULL;
+
+#ifdef MS_WINDOWS
+        dup_fd = _dup(fd);
+        if (dup_fd != -1) {
+            f = _fdopen(dup_fd, "r");
+        }
+        if (!f) {
+            _close(dup_fd);
+        }
+#else
+        dup_fd = dup(fd);
+        if (dup_fd != -1) {
+            f = fdopen(dup_fd, "r");
+        }
+        if (!f) {
+            close(dup_fd);
+        }
+#endif
+        if (!f) {
+            PyErr_SetFromErrno(PyExc_OSError);
+        } else {
+            PyRun_AnyFileEx(f, path, 1);
+        }
+
+        if (PyErr_Occurred()) {
+            PyErr_FormatUnraisable("Error executing debugger script %s", path);
+        }
+    }
+
+    PyObject* res = PyObject_CallMethod(fileobj, "close", "");
```

Comment:
Is it a noarg method? if so, maybe use `PyObject_CallMethodNoArgs(fileobj, &_Py_ID(close))`. And if `PyObject_AsFileDescriptor` failed, shouldn't we first clear the exception and attempt to close?

Reply:
No, we want it to appear as an unraisable error by PyErr_FormatUnraisable. We then close the file as we exit the fd == -1 branch

Reply:
I see that `PyErr_FormatUnraisable` is actually clearing the error indicator, so indeed `PyObject_CallMethodNoArgs` should not actually complain about it to be set (am I right?)

Reply:
Correct. Or at least that's my current understanding and the original intent